### PR TITLE
Unique Face Object Tracking

### PR DESCRIPTION
**Pull request by tobitech**
### **User description**
Story: https://app.shortcut.com/smileid/story/xxx

## Summary

A few sentences/bullet points about the changes

## Known Issues

Any shortcomings in your work. This may include corner cases not correctly handled or issues related
to but not within the scope of your PR. Design compromises should be discussed here if they were not
already discussed above.

## Test Instructions

Concise test instructions on how to verify that your feature works as intended. This should include
changes to the development environment (if applicable) and all commands needed to run your work.

## Screenshot

If applicable (e.g. UI changes), add screenshots to help explain your work.

___

### **PR Type**
Enhancement

___

### **Description**
• Implement face tracking functionality with dedicated tracking manager
• Integrate tracking state management in selfie capture workflows
• Add tracking reset capabilities for improved user experience
• Enable face tracking in enhanced selfie capture by default

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>EnhancedFaceDetector.swift</strong><dd><code>Integrate face tracking capabilities into enhanced face detector</code></dd></summary>
<hr>

Sources/SmileID/Classes/FaceDetector/EnhancedFaceDetector.swift

• Add face tracking integration with <code>FaceTrackingManager</code><br> • Implement <br>tracking-enabled and tracking-disabled processing paths<br> • Add delegate <br>methods for tracking state changes and resets<br> • Refactor face <br>detection logic to support tracking workflow

</details>

  </td>
  <td><a href="https://github.com/smileidentity/ios/pull/361/files#diff-2651d30e57c28a4098bd77dd8dec05052d22ac88047a605a3e11322dd3a029d2">+131/-1</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>FaceTrackingManager.swift</strong><dd><code>Add dedicated face tracking manager implementation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Sources/SmileID/Classes/FaceDetector/FaceTrackingManager.swift

• Create new face tracking manager with Vision framework integration<br> • <br>Implement tracking state management and confidence validation<br> • Add <br>queue-based frame processing for performance optimization<br> • Handle <br>tracking loss and recovery scenarios

</details>

  </td>
  <td><a href="https://github.com/smileidentity/ios/pull/361/files#diff-d327886ce122d7339a728657bff9f66ed7c503beef748ac92a9ba59534ea954a">+140/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>FaceTrackingState.swift</strong><dd><code>Define face tracking state management types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Sources/SmileID/Classes/FaceDetector/FaceTrackingState.swift

• Define face tracking states (detecting, tracking, lost, reset)<br> • <br>Create tracking error types and delegate protocol<br> • Add configuration <br>struct with default tracking parameters

</details>

  </td>
  <td><a href="https://github.com/smileidentity/ios/pull/361/files#diff-8090bc11c9fbe74e8ce46f0ea703bdcdb4acfb1cfe83b30c8ad5c77d9ce4bf11">+37/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>LivenessCheckManager.swift</strong><dd><code>Expose liveness challenge reset method</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Sources/SmileID/Classes/FaceDetector/LivenessCheckManager.swift

• Change `resetLivenessChallenge` method visibility to public

</details>

  </td>
  <td><a href="https://github.com/smileidentity/ios/pull/361/files#diff-605f8269c788f1b636703fba214d87fd65a3ee9bd9aed81b22280ace81f8f1d0">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>EnhancedSmartSelfieViewModel.swift</strong><dd><code>Integrate face tracking in enhanced selfie view model</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Sources/SmileID/Classes/SelfieCapture/EnhancedSmartSelfieViewModel.swift

• Enable face tracking by default in enhanced selfie capture<br> • <br>Implement tracking state change handlers with user instructions<br> • Add <br>tracking reset handling with liveness challenge reset<br> • Reset selfie <br>capture state when face tracking resets

</details>

  </td>
  <td><a href="https://github.com/smileidentity/ios/pull/361/files#diff-72f9bb5deaeb7c97fd27a481ca9e4fdf2352d7e6dd9413c3e08c0e9df4b8aa90">+35/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>SelfieViewModel.swift</strong><dd><code>Integrate face tracking in standard selfie view model</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Sources/SmileID/Classes/SelfieCapture/SelfieViewModel.swift

• Add face tracking manager integration and delegate implementation<br> • <br>Implement tracking state change handling with directive updates<br> • Add <br>face tracking reset functionality for new face detection<br> • Handle <br>tracking errors with appropriate user feedback

</details>

  </td>
  <td><a href="https://github.com/smileidentity/ios/pull/361/files#diff-ce09d5beb9fce2c0797ee811714d6123d4694f641aec1bdaa5cfa9447eee3369">+66/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

File: Sources/SmileID/Classes/FaceDetector/EnhancedFaceDetector.swift [hunk 1]
```diff
@@ -22,24 +22,64 @@ protocol FaceDetectorResultDelegate: AnyObject {
...
+  )
 }
 
 class EnhancedFaceDetector: NSObject {
   private let cropSize = (width: 120, height: 120)
   private let faceMovementThreshold: CGFloat = 0.15
 
   private var sequenceHandler = VNSequenceRequestHandler()
+  private var faceTrackingManager: FaceTrackingManager
+  private var isTrackingEnabled: Bool = false
```
Comment (jumaallan) [inline Sources/SmileID/Classes/FaceDetector/EnhancedFaceDetector.swift:44]: so this variable can be removed

File: Sources/SmileID/Classes/FaceDetector/EnhancedFaceDetector.swift [hunk 2]
```diff
@@ -22,24 +22,64 @@ protocol FaceDetectorResultDelegate: AnyObject {
...
+  func enableFaceTracking(_ enabled: Bool) {
+    isTrackingEnabled = enabled
+    if !enabled {
+      faceTrackingManager.resetTracking()
+    }
   }
 
   /// Run Face Capture quality and Face Bounding Box and roll/pitch/yaw tracking
   func processImageBuffer(_ imageBuffer: CVPixelBuffer) {
+    if isTrackingEnabled {
```
[the remark this answers is not in the document]
Comment (tobitech) [reply Sources/SmileID/Classes/FaceDetector/EnhancedFaceDetector.swift:64]: I put this there to be able to compare the two scenarios while testing. I will remove it before merging.

[the remark this answers is not in the document]
Comment (jumaallan) [reply Sources/SmileID/Classes/FaceDetector/EnhancedFaceDetector.swift:64]: ooh okay

**Comment by tobitech** [comment]
Will open a new PR with the same changes but on the new project setup.
